### ISSUE TITLE
Support Windows Runtime async callbacks

### ISSUE THREAD

**Issue by gspencergoog** [dart-lang/sdk#45342]
from @timsneath:
> The modern Windows Runtime APIs are mostly asynchronous. For example, the modern file picker API is [PickSingleFileAsync](https://docs.microsoft.com/en-us/uwp/api/windows.storage.pickers.fileopenpicker.picksinglefileasync?view=winrt-19041), which requires a delegate to be called upon completion. It’s unclear whether Dart today provides suitable primitives to allow the required thread synchronization after a callback.
>
>Aside from the known FFI work that is already underway [...] this is the only remaining obstacle as far as I’m aware to completing a full projection of the Windows API to Dart. 

Would it be possible for Dart to expose/provide (or if they already exist, just explain) the primitives needed to be able to enable this use case? It would make Dart programs (including Flutter apps) on Windows quite powerful, as they could take advantage of the entire Windows API directly.

**Comment by mit-mit** [comment]
cc @dcharkes if he know about this

**Comment by dcharkes** [comment]
Currently, `dart:ffi` does not support callbacks on an arbitrary thread (https://github.com/dart-lang/sdk/issues/37022), which you would use to get a callback from C when it completes, nor asynchronous calls (https://github.com/dart-lang/sdk/issues/35766), which you would use to do a blocking C call for the result but let Dart code execute in the mean time.

With reading the API docs, the thing that you could achieve right now with `dart:ffi` without writing any C code, is binding to the methods on `IAsyncOperation` and polling `Status` with sleeps in the mean time, and when that is done call `GetResult` (which is blocking until it's completed). This could then be exposed with a `Future` in Dart.

After addressing https://github.com/dart-lang/sdk/issues/35766, you should be able to do the FFI call to `GetResult` which would then essentially be an await and would yield the Dart execution thread to other Dart code rather than blocking all Dart execution and return a future.

@gspencergoog We're already tracking async support for `dart:ffi`. Does the polling workaround work for you in the mean time?

(Reassigning labels, assuming this is in the context of `dart:ffi` and `package:win32` rather than `dart:io`.)

**Comment by gspencergoog** [comment]
Could we also wrap `IAsyncOperationWithProgress<TResult,TProgress>` in a similar way?  Then we could cover the majority of all async Windows operations.

I'm not too psyched about polling the status with sleeps, but if it were wrapped well enough and could be upgraded later to use tasks or something without changing the wrapped API, that would be fine as a start.

(@timsneath would this fit your requirements?)

**Comment by timsneath** [comment]
Yes, I think so! This is stretching to the edge of my knowledge / expertise
(there's a reason I'm not a Google SWE!), but I'd be more than happy to
pair up with someone to try and get this working.

**Comment by dcharkes** [comment]
@gspencergoog if you're able to load an extra DLL, you can use the native ports mechanism for callbacks ([sample](https://github.com/dart-lang/sdk/blob/master/samples/ffi/async/sample_native_port_call.dart)). You'd write a C function that send a message, pass that function pointer to the win32 API, and receive the message in Dart and expose that as a Future again.

Of course, that involves writing a small C library and building it - which is not in the spirit of package:win32 being Dart-code-only. Where do you want to use this?

**Comment by timsneath** [comment]
My current progress is here, for anyone interested in experimenting. I put this down a while back to focus on other areas of the package, but this is starting to become the next big hill to climb! 
  https://github.com/timsneath/win32/issues/97
